### DISCUSSION TITLE
Visualizing activations for TensorflowPredictMusiCNN

### DISCUSSION THREAD

**Discussion by bclark288** [MTG/essentia#1264] [Essentia Models]
Hi, I'm developing an application that would allow a musician to view the activations from TensorflowPredictMusiCNN (using msd-musicnn-1.pb) superimposed on the waveform.  However, I'm having trouble calculating the time axis for the activations.  I calculated patch length as the length of the audio in samples / 16000 / number of activations.  This number comes out to be somewhere around 1.5 seconds (it's not exactly the same for every song).  However, when I sum the durations for all the activations, the duration of the song is significantly different that the actual duration of the audio.  Can you help me understand what I am missing?

**Comment by Hikari-Tsai** [comment]
Do not derive the activation spacing by dividing the clip duration by the number of activations. MusiCNN uses fixed analysis geometry.

In Essentia's `TensorflowPredictMusiCNN` implementation:

- audio sample rate: 16,000 Hz
- analysis frame: 512 samples
- frame hop: 256 samples = 16 ms
- patch size: 187 frames (about 2.992 s)
- default patch hop: 93 frames = 1.488 s

Therefore activation/patch `i` starts at approximately:

```python
start_seconds = i * 93 * 256 / 16000
center_seconds = start_seconds + (187 * 256 / 16000) / 2
```

Use `start_seconds` if the visualization represents the patch's beginning, or `center_seconds` if you want the prediction centered over its receptive field.

The summed patch durations do not equal the song duration because adjacent patches overlap by roughly 50%. In addition, the handling of the last incomplete patch (`lastPatchMode`) determines whether the tail is discarded, repeated, or padded. The exact span is approximately `patch_duration + (N - 1) * patch_hop_duration`, not `N * patch_duration`.

If you configure a different `patchHopSize` or `patchSize`, substitute those values in the formula rather than inferring them from the output count.
